File: java/636.exclusive-time-of-functions.java
class Solution {
    public int[] exclusiveTime(int n, List<String> logs) {
        if (n < 1 || logs == null) {
            return null;
        }
        int[] result = new int[n];
        // print(result);
        int prev = 0;
        int curr = 0;
        Deque<Integer> stack = new LinkedList<>();
        for (String log : logs) {
            String[] cmds = log.split(":");
            int id = Integer.parseInt(cmds[0]);
            String action = cmds[1];
            int endTime = Integer.parseInt(cmds[2]);
            // handler log
            if (action.equals("start")) {
                if (!stack.isEmpty()) {
                    result[stack.peekFirst()] += endTime - prev;
                    prev = endTime;
                }
                stack.offerFirst(id);
            } else {
                result[stack.pollFirst()] += endTime - prev + 1;
                prev = endTime + 1;
            }
            // print(result);
        }
        return result;
    }
}

File: python3/636.exclusive-time-of-functions.py
class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        ans = [0] * n
        
        events = []
        
        for log in logs:
            args = log.split(':')
            events.append(
                (
                    int(args[0]), 
                    args[1], 
                    # make 'start' and 'end' in same time scale.
                    int(args[2]) + int(args[1] == 'end')
                )
            )
            
        stack = []
        
        last_time = 0
        
        for id, type, timestamp in events:
            
            if type == 'start':
                if stack:
                    ans[stack[-1]] += (timestamp - last_time)
                stack.append(id)
                last_time = timestamp
                
            elif type == 'end':
                ans[stack[-1]] += (timestamp - last_time)
                stack.pop()
                
            last_time = timestamp
                
            
            
        return ans
                
        
